### PR TITLE
Do not delete polymorphic objects without a virtual destructor

### PR DESCRIPTION
**Pull request by omajid**
SEI CERT C++ Coding Standard says:

> Do not delete an object of derived class type through a pointer to its
> base class type that has a non-virtual destructor. Instead, the base
> class should be defined with a virtual destructor. Deleting an object
> through a pointer to a type without a virtual destructor results in
> undefined behavior.

See https://wiki.sei.cmu.edu/confluence/display/cplusplus/OOP52-CPP.+Do+not+delete+a+polymorphic+object+without+a+virtual+destructor

Clang generally warns about this, but we disabled the warning via
-Wno-delete-non-virtual-dtor. This commit re-enables the warning and
fixes up all the code that hits the warning.

I do not know the performance impact of this change. Any tips on what I can do to measure that?

### CODE REVIEW

**Comment by mikedn** [comment]
> I do not know the performance impact of this change. Any tips on what I can do to measure that?

There's practically nothing in this change that could have any meaningful performance impact.

File: src/ToolBox/SOS/lldbplugin/inc/lldbservices.h [hunk 1]
```diff
@@ -237,6 +237,9 @@ MIDL_INTERFACE("2E6C569A-9E14-4DA4-9DFC-CDB73A532566")
 ILLDBServices : public IUnknown
 {
 public:
+
+    virtual ~ILLDBServices() = default;
```
Comment (mikedn) [inline src/ToolBox/SOS/lldbplugin/inc/lldbservices.h]: Seems kind of odd to have a destructor in a COM interface. Well, this is probably not real COM but still, it inherits `IUnknown`. Is there actually code that uses delete on a ILLDBService pointer?

Comment (omajid) [reply src/ToolBox/SOS/lldbplugin/inc/lldbservices.h]: Here's the error from clang:

```
/home/omajid/devel/dotnet/coreclr/src/ToolBox/SOS/lldbplugin/services.cpp:68:9: error: delete called on non-final    
      'LLDBServices' that has virtual functions but non-virtual destructor [-Werror,-Wdelete-non-virtual-dtor]       
        delete this;
        ^
1 error generated.
```

Comment (mikedn) [reply src/ToolBox/SOS/lldbplugin/inc/lldbservices.h]: I think that the proper solution here is to make `LLDBServices`'s destructor virtual or make the class final. Nobody's going to use `delete` on a COM interface (there's no `ILLDBService* p = … ; delete p;` anywhere in the code)

File: src/inc/iceefilegen.h [hunk 2]
```diff
@@ -76,6 +76,8 @@ typedef HRESULT (__stdcall * PFN_DestroyICeeFileGen)(ICeeFileGen ** ceeFileGen);
 
 class ICeeFileGen {
   public:
+    virtual ~ICeeFileGen() = default;
```
Comment (mikedn) [inline src/inc/iceefilegen.h]: This probably has the same issue as `LLDBServices`, there's probably a class that implements this interface and that class needs a virtual destructor.

Comment (mikedn) [reply src/inc/iceefilegen.h]: Ah, I see. This isn't actually a COM interface, despite the name. So I suppose this one is fine.

File: src/ToolBox/SOS/lldbplugin/inc/lldbservices.h (hunk 1 above)
Comment (omajid) [reply src/ToolBox/SOS/lldbplugin/inc/lldbservices.h]: I totally mixed up `ILLDBService` and `LLDBService` :( Fixed now.

**Comment by omajid** [comment]
@janvorli I have rebased this commit onto master. If the changes still look okay, can you please merge them?

**Comment by janvorli** [comment]
@omajid it looks good, thank you!

**Comment by omajid** [comment]
Thanks for merging this!

**Pull request merged**